Editor: Focus value editor on type change in Dictionary and Array editors

EmrysMyrddin opened the pull request:
In the inspector, when editing an exported `Dictionary` property, the focus is lost once a type is chosen for the new key, the new value or an existing value.


https://github.com/godotengine/godot/assets/3855602/4055f3f5-cd4e-4fd6-a839-a19de3051806

This PR is fixing this by focusing the editor property after changing the type:

https://github.com/godotengine/godot/assets/3855602/f51e19b4-3706-4dbc-8f73-4ee6fa10c25d

EmrysMyrddin committed:
Editor: Focus value editor on type change in Dictionary and Array editors

AThousandShips commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -1163,6 +1164,9 @@ void EditorPropertyDictionary::update_property() {
 			prop->connect("property_changed", callable_mp(this, &EditorPropertyDictionary::_property_changed));
 			prop->connect("object_id_selected", callable_mp(this, &EditorPropertyDictionary::_object_id_selected));
+			if (change_index == changing_type_index) {
+				callable_mp(prop, &EditorProperty::grab_focus).bind(0).call_deferred();
```
```suggestion
				callable_mp(prop, &EditorProperty::grab_focus).call_deferred(0);
```

KoBeWi commented:
The focus is still lost when pressing the Add Key/Value Pair button. Any chance this could also be improved?

ajreckof commented:
This PR was superseded by #88231. But I feel like the idea of switching the focus from the change_type button to the EditorProperty is a good idea and could be a great improvement.

EmrysMyrddin commented:
@ajreckof Thank you for the update, should I continue to work on this then ?

ajreckof commented:
I think the switch of focus would be a great improvement and I would 100% be for it.  And I don't see any reasons it wouldn't be accepted. If you need any help with it feel free to send me a message on rocket chat.

EmrysMyrddin commented:
Wow, @ajreckof great work on this refactor! The code is muuuch clearer now, it's even cleaner and more understandable to fix the focus now!

I've rebased and fixed issue pointed out by @KoBeWi, the problem was that the focus was taken only on property update, which doesn't happens if you open the change type popup but chose the same type already selected.

EmrysMyrddin commented:
After looking at @ajreckof refactoring, I understood that the Array Editor works pretty much the same, so I also made the change for it, so that both Dictionary and Array editors have the same behaviour.

KoBeWi commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -252,6 +252,11 @@ void EditorPropertyArray::_change_type_menu(int p_index) {
 	}
 
+	if (changing_type_index == EditorPropertyArrayObject::NOT_CHANGING_TYPE) {
+		ERR_PRINT("Tried to change type of an array item, but no item was selected.");
+		return;
```
```suggestion
		ERR_FAIL_MSG("Tried to change type of an array item, but no item was selected.");
```

AThousandShips commented:
Please squash your commits into one, see [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase)

EmrysMyrddin commented:
Sure, done ✅

KoBeWi commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -255,6 +255,11 @@ void EditorPropertyArray::_change_type_menu(int p_index) {
 
+	if (changing_type_index == EditorPropertyArrayObject::NOT_CHANGING_TYPE) {
+		ERR_FAIL_MSG("Tried to change type of an array item, but no item was selected.");
+		return;
```
`ERR_FAIL` already returns, so the `return` should be removed.

KoBeWi commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -934,6 +943,10 @@ void EditorPropertyDictionary::_change_type_menu(int p_index) {
 			break;
 
+		case EditorPropertyDictionaryObject::NOT_CHANGING_TYPE:
+			ERR_FAIL_MSG("Tried to change the type of a dict key or value, but nothing was selected.");
```
Sorry, I didn't notice this one uses `break` 🙃
It should be `ERR_PRINT` like before.

EmrysMyrddin replied:
Aaaah ok I just understood what this things are doing ><' I'm really not a C++ programmer, not used to this macros ^^'

EmrysMyrddin replied:
Should I use the variant that also notifies the editor ?

EmrysMyrddin replied:
From what I understood in another PR, `ERR_PRINT` only prints to the standard output, while `ERR_PRINT_ED` is also displaying a nice toast notification inside the editor (at bottom right corner by default if I well remember)

EmrysMyrddin replied:
This should never happen, I don't think it's really useful to display a notification for this.

EmrysMyrddin commented:
Should keep the branch as much up to date as possible by rebasing on master on a regular bases ?

KoBeWi commented:
Only when there are conflicts or a long time has passed (i.e months). Usually it's not necessary.

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

Calinou reviewed (commented):
Tested locally, it works as expected. This is a nice usability boost for keyboard usage :slightly_smiling_face: 

(Sorry, forgot to refresh before posting the review.)
